### PR TITLE
Moving diagnostics into a service

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Moving diagnostics into a service

#### Ticket Link
[MM-26198](https://mattermost.atlassian.net/browse/MM-26198)

### CODE REVIEW

File: services/diagnostics/diagnostics_test.go
Comment (hanzei) [inline services/diagnostics/diagnostics_test.go:4]: One thought I wanted to share: Outside of the code in e.g. the documentations this feature is called `telemetry`. There have been developers getting confused by this naming inconsistency.

Maybe this PR is a good opportunity to rename everything to `telemetry`?

Comment (jespino) [reply services/diagnostics/diagnostics_test.go:4]: Very good point, I think you are totally right

File: services/telemetry/telemetry.go
Comment (reflog) [inline services/telemetry/telemetry.go:1167]: why is there still segment related code? I believe it's already replaced bu rudder 🤔

Comment (jespino) [reply services/telemetry/telemetry.go:1167]: I don't know, that is a question for @grundleborg

File: services/telemetry/telemetry.go
Comment (isacikgoz) [inline services/telemetry/telemetry.go:111]: Shouldn't we return an error here? The method logic seemed to be have different paths on different circumstances. At least might want to log at debug when calling this method.

File: services/telemetry/telemetry_test.go
Comment (isacikgoz) [inline services/telemetry/telemetry_test.go:299]: Should we leave these commented out? There are similar occurrences in the test code.

Comment (jespino) [reply services/telemetry/telemetry_test.go:299]: Actually that was uncommented before and I had to comment it to not mock all the plugins environment object.

File: services/telemetry/telemetry.go
Comment (jespino) [reply services/telemetry/telemetry.go:111]: This was the previous existing behavior but I think is a good thing to at least, log it.

Comment (jespino) [reply services/telemetry/telemetry.go:111]: Done

File: services/telemetry/telemetry.go
Comment (grundleborg) [reply services/telemetry/telemetry.go:1168]: Putting a do not merge label on this PR to block it until https://github.com/mattermost/mattermost-server/pull/14712 is through QA and merged, as it is considerably higher priority than this (sorry Jesus).

Comment (jespino) [reply services/telemetry/telemetry.go:1168]: Ohh, sure, first things first :)

File: model/migration.go
Comment (grundleborg) [inline model/migration.go:7]: Why are you moving this to model? It's not part of the migrations model but a special "migration" that exists entirely in the app layer.

Comment (grundleborg) [reply model/migration.go:7]: Also has nothing to do with telemetry.

Comment (jespino) [reply model/migration.go:7]: This is used as part of the telemetry here https://github.com/mattermost/mattermost-server/pull/14832/files#diff-6d075395bdd93d9b78959b73d7a2f3d9R931 so we have 2 options, or use directly the string in the telemetry to get key, or expose this outside of the app to avoid a cyclic dependency.

Other option is to remove that data from telemetry (Probably we have some data that we can remove from telemetry because is no longer interesting).

Comment (grundleborg) [reply model/migration.go:7]: Gross, but 0/5 and not really your fault - it's just a somewhat inevitable side effect of having that app layer migration in the first place...

File: config/client.go
Comment (grundleborg) [inline config/client.go:251]: Can we also add an additional field called `telemetryId` with the same value, so we can migrate gradually on the clients side.

File: services/telemetry/telemetry.go
Comment (grundleborg) [inline services/telemetry/telemetry.go:106]: Not an important point at all, but why `ds`?

Comment (jespino) [reply services/telemetry/telemetry.go:106]: Good question, it was `DiagnosticsService` before, not makes a lot of sense to change everything to `ts`.

**Comment by grundleborg** [comment]
@jespino I've got a ticket coming up in the next couple of weeks that will probably conflict really badly with this one - do you want to try and get this finished up and merged before I start on that? Are there any changes still pending on this ticket, or should I go ahead and review it again now?

**Comment by jespino** [comment]
@grundleborg this was waiting for another review and then for QA, I'm perfectly fine merging master here, and fix all the conflicts the next Wednesday. If we going to make drastic changes to diagnostics code, probably we should do this changes before, to easy develop/test/iterate the new feature. But It is up to you. I don't mind to discard this PR and do it all over again after your feature.

**Comment by grundleborg** [comment]
I would prefer to merge this PR before I start work if it's ready. I'll review it later today.

**Comment by jespino** [comment]
@grundleborg It was ready, but probably is better to review this after I merge master on it.

**Comment by jespino** [comment]
@grundleborg I merged master, but probably I need to compare the changes made to de code by myself too.

**Comment by grundleborg** [comment]
@jespino ok well give me a ping when you've had a chance to do that and I will review afterwards.

**Comment by jespino** [comment]
@grundleborg I think now is ready

**Review by grundleborg — APPROVED** [review]
LGTM

**Review by lindalumitchell — APPROVED** [review]

**Pull request merged**
